test/common/http/http1/codec_impl_test.cc

```diff
@@ -728,6 +728,35 @@ TEST_F(Http1ClientConnectionImplTest, WatermarkTest) {
       ->onUnderlyingConnectionBelowWriteBufferLowWatermark();
 }
 
+// For issue #1421 regression test that Envoy's HTTP parser applies header limits early.
+TEST_F(Http1ServerConnectionImplTest, TestCodecHeaderLimits) {
+  initialize();
+
+  std::string exception_reason;
+  NiceMock<Http::MockStreamDecoder> decoder;
+  Http::StreamEncoder* response_encoder = nullptr;
+  EXPECT_CALL(callbacks_, newStream(_))
+      .WillOnce(Invoke([&](Http::StreamEncoder& encoder) -> Http::StreamDecoder& {
+        response_encoder = &encoder;
+        return decoder;
+      }));
+
+  Buffer::OwnedImpl buffer("GET / HTTP/1.1\r\n");
+  codec_->dispatch(buffer);
+  std::string long_string = "foo: " + std::string(1024, 'q') + "\r\n";
+  for (int i = 0; i < 79; ++i) {
+    buffer = Buffer::OwnedImpl(long_string);
+    codec_->dispatch(buffer);
+  }
+  try {
+    buffer = Buffer::OwnedImpl(long_string);
+    codec_->dispatch(buffer);
+  } catch (const Envoy::EnvoyException& ex) {
+    exception_reason = ex.what();
```

Comment:
What happens on the data path when we take an exception? Most exceptions in Envoy that I'm familiar with today happen on the config update path.

Reply:
Protocol errors are raised as exceptions.

Reply:
P.S. You can use EXPECT_THROW_WITH_MESSAGE here.

Reply:
Ah, I didn't know about that one, thanks!
(I'm totally claiming the macro means I haven't written exception code yet.  heh heh heh)